### PR TITLE
fix (Resources/useDateTimePicker): Some months are missing ,some days are duplicates in the calendar

### PR DESCRIPTION
**Pull request by Noha-ElAbrouki**
refs: 
https://centreon.atlassian.net/jira/software/c/projects/MON/boards/106?modal=detail&selectedIssue=MON-15805&assignee=61d809797c6f9800706f84a7
https://centreon.atlassian.net/jira/software/c/projects/MON/boards/106?modal=detail&selectedIssue=MON-15805&assignee=61d809797c6f9800706f84a7
video :(utc, london,paris,sydney)

https://user-images.githubusercontent.com/97687698/203812616-3c40a89d-eb72-4404-9531-db426ec63ee5.mp4

### CODE REVIEW

File: centreon/www/front_src/src/Resources/useDateTimePickerAdapter.ts [hunk 1]
```diff
@@ -196,6 +214,45 @@ const useDateTimePickerAdapter = (): UseDateTimePickerAdapterProps => {
...
+      const nestedWeeks: Array<Array<dayjs.Dayjs>> = [];
+
+      while (current.isBefore(end)) {
+        const weekNumber = Math.floor(count / 7);
+        nestedWeeks[weekNumber] = nestedWeeks[weekNumber] || [];
+        nestedWeeks[weekNumber].push(current);
+
+        current = current.tz(timezone, true).add(1, 'day') as dayjs.Dayjs;
+        count += 1;
+      }
```
Comment (bdauria) [inline centreon/www/front_src/src/Resources/useDateTimePickerAdapter.ts:251]: - Avoid mutation (use a functional approach, using rambda for instance)
- Try to reuse a function when the code is similar to improve reusability

**Review by bdauria — COMMENTED** [review]
@Noha-ElAbrouki Some visual tests using different key timezones and edge cases would be appreciated

**Comment by Thebarda** [comment]
Is it possible to add some tests?

File: centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  'Tuesday',
+  'Wednesday',
+  'Thursday',
+  'Friday',
+  'Saturday',
+];
+
+const date = '2023-02-01T12:59:41.041Z';
+const timeZoneParis = 'Europe/Paris';
+const Months2023 = [
```
Comment (Thebarda) [inline centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx:36]:
```suggestion
const months2023 = [
```

File: centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+import dayjs from 'dayjs';
+import localizedFormatPlugin from 'dayjs/plugin/localizedFormat';
+import timezonePlugin from 'dayjs/plugin/timezone';
+import utcPlugin from 'dayjs/plugin/utc';
+import { useAtomValue } from 'jotai';
+import { equals } from 'ramda';
+
+import { LocalizationProvider } from '@mui/x-date-pickers';
+
+import { userAtom } from '@centreon/ui-context/src';
```
Comment (Thebarda) [inline centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx:13]:
```suggestion
import { userAtom } from '@centreon/ui-context';
```

File: centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+      userData.result.current.timezone = timeZoneParis;
+      userData.result.current.locale = 'en_US';
+    });
+  });
+  it('first test ,tz = europe/Paris', () => {
+    const { result } = renderHook(() => useDateTimePickerAdapter());
+
+    act(() => {
+      const { Adapter } = result.current;
+      cy.mount(
```
Comment (Thebarda) [inline centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx:68]: Use the mount custom command: https://github.com/centreon/centreon/blob/develop/centreon/cypress/support/component.tsx

File: centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    const currentMonth = currentDateByTimeZone.format('M');
+    const month = Months2023[Number(currentMonth) - 1];
+    const { firstDay, lastDay, numberWeeks } = Object.values(month)[0];
+
+    cy.get('input').should('have.value', dateInput);
+    cy.get('input').click();
+    cy.get('[role="rowgroup"]')
+      .children()
+      .each(($week, indexWeek, $listWeeks) => {
+        if (indexWeek === 0) {
```
Comment (Thebarda) [inline centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx:92]:
```suggestion
        if (equals(indexWeek, 0)) {
```

File: centreon/www/front_src/src/Resources/useDateTimePickerAdapter.ts [hunk 6]
```diff
@@ -196,6 +230,59 @@ const useDateTimePickerAdapter = (): UseDateTimePickerAdapterProps => {
       );
     };
 
+    public chunk = ({ array, size }: Chunk): Array<unknown> => {
```
Comment (Thebarda) [inline centreon/www/front_src/src/Resources/useDateTimePickerAdapter.ts:233]:
```suggestion
    public getChunkFromArray = ({ array, size }: Chunk): Array<unknown> => {
```

File: centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+          />
+        </LocalizationProvider>,
+      );
+    });
+
+    const currentDateByTimeZone = dayjs(date).tz(timeZoneParis);
+
+    const dateInput = currentDateByTimeZone.format('L hh:mm A');
+    const currentMonth = currentDateByTimeZone.format('M');
+    const month = Months2023[Number(currentMonth) - 1];
```
Comment (Thebarda) [inline centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx:84]:
```suggestion
    const month = months2023[Number(currentMonth) - 1];
```

File: centreon/www/front_src/src/Resources/useDateTimePickerAdapter.ts [hunk 8]
```diff
@@ -196,6 +230,59 @@ const useDateTimePickerAdapter = (): UseDateTimePickerAdapterProps => {
...
+            return [...acc, newCurrent];
+          }
+          const newCurrent = acc[currentIndex].add(1, 'day');
+
+          return [...acc, newCurrent];
+        },
+        [current],
+      );
+
+      const arrayWeeks = this.chunk({
```
Comment (Thebarda) [inline centreon/www/front_src/src/Resources/useDateTimePickerAdapter.ts:278]:
```suggestion
      const arrayWeeks = this.getChunkFromArray({
```

File: centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/DateTimePickerInput.tsx [hunk 9]
```diff
@@ -80,6 +80,7 @@ const DateTimePickerInput = ({
       minDate={minDate && dayjs(minDate)}
       open={isOpen}
       renderInput={renderDateTimePickerTextField(blur)}
+      showToolbar={false}
```
[the remark this answers is not in the document]
Comment (Noha-ElAbrouki) [reply centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/DateTimePickerInput.tsx:83]: by default is not displayed , but when i run cypress tests it took the value true o.O idk :s . so i put the value to false to have the same calendar in tests with web

File: centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+describe('calendar', () => {
+  beforeEach(() => {
+    const userData = renderHook(() => useAtomValue(userAtom));
+
+    act(() => {
+      userData.result.current.timezone = timeZoneParis;
+      userData.result.current.locale = 'en_US';
+    });
+  });
+  it('first test ,tz = europe/Paris', () => {
```
Comment (bdauria) [inline centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx:63]: This test name does not sound relevant to me

File: centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+const setStart = undefined;
+describe('calendar', () => {
+  beforeEach(() => {
+    const userData = renderHook(() => useAtomValue(userAtom));
+
+    act(() => {
+      userData.result.current.timezone = timeZoneParis;
+      userData.result.current.locale = 'en_US';
+    });
+  });
```
Comment (bdauria) [inline centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx:62]:
```suggestion
  });
  
```

File: centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+    February: {
+      firstDay: { day: days[3], indexDayInRowWeek: 3, value: 1 },
+      lastDay: { day: days[2], indexDayInRowWeek: 2, value: 28 },
+      numberWeeks: 5,
+    },
+  },
+];
+
+const changeDate = (): void => undefined;
+const setStart = undefined;
```
Comment (bdauria) [inline centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx:53]:
```suggestion

const setStart = undefined;
```

File: centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx (hunk 4 above)
Comment (Noha-ElAbrouki) [reply centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx:70]: it's the custom command , i didn't import mount from cypress 🤔

File: centreon/www/front_src/src/Resources/useDateTimePickerAdapter.ts [hunk 13]
```diff
@@ -196,6 +230,71 @@ const useDateTimePickerAdapter = (): UseDateTimePickerAdapterProps => {
...
+        : customStart;
+      const currentEnd = end.isUTC() ? end.tz(timezone, true) : customEnd;
+
+      const numberOfDaysInCurrentMonth = currentEnd.diff(
+        currentStart,
+        'd',
+        true,
+      );
+
+      const daysOfMonthWithTimeZone = [
```
Comment (Thebarda) [inline centreon/www/front_src/src/Resources/useDateTimePickerAdapter.ts:271]:
```suggestion
      const daysOfMonthWithTimezone = [
```

File: centreon/www/front_src/src/Resources/useDateTimePickerAdapter.ts [hunk 14]
```diff
@@ -196,6 +230,71 @@ const useDateTimePickerAdapter = (): UseDateTimePickerAdapterProps => {
...
+        (acc, _, currentIndex) => {
+          if (acc[currentIndex].isUTC()) {
+            const newCurrent = acc[currentIndex]
+              .utc()
+              .add(1, 'day')
+              .tz(timezone, true);
+
+            return [...acc, newCurrent];
+          }
+          const newCurrent = acc[currentIndex].add(1, 'day');
```
Comment (Thebarda) [inline centreon/www/front_src/src/Resources/useDateTimePickerAdapter.ts:283]:
```suggestion

          const newCurrent = acc[currentIndex].add(1, 'day');
```

**Comment by bdauria** [comment]
@Noha-ElAbrouki I suggest we only test the Europe/Paris TZ for now by using tags. We will create separate pipeline runs afterwards to execute them less frequently

**Review by Thebarda — COMMENTED** [review]
Run the pnpm eslint:fix in order to fix eslint errors

File: centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx [hunk 15]
```diff
@@ -0,0 +1,496 @@
...
+    December: {
+      date: '2023-12-01T12:59:41.041Z',
+      firstDay: { day: days[5], indexDayInRowWeek: 5, value: 1 },
+      lastDay: { day: days[0], indexDayInRowWeek: 0, value: 31 },
+      numberWeeks: 6,
+    },
+  },
+];
+
+const month2023Reverse = [
```
[the remark this answers is not in the document]
Comment (Thebarda) [reply centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx:136]: Oh I didn't notice that

File: centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx [hunk 16]
```diff
@@ -0,0 +1,471 @@
...
+
+          return expect(isDuplicateExist).to.be.false;
+        });
+
+        const currentMonth = Object.keys(element)[0];
+        getPreviousNextMonth({ currentMonth, labelButton: item.button });
+      });
+    });
+
+    it.only(`checks that the first day must match with the start of the week when the ${item.button} button is clicked`, () => {
```
Comment (Thebarda) [inline centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx:363]:
```suggestion
    it(`checks that the first day must match with the start of the week when the ${item.button} button is clicked`, () => {
```

File: centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx [hunk 17]
```diff
@@ -0,0 +1,471 @@
...
+        cy.get('@listDaysInFirstWeek')
+          .eq(firstDay.indexDayInRowWeek)
+          .as('firstDayInWeek');
+        cy.get('@firstDayInWeek').contains(firstDay.value);
+
+        const currentMonth = Object.keys(element)[0];
+        getPreviousNextMonth({ currentMonth, labelButton: item.button });
+      });
+    });
+    it.only(`checks that the last day must match with the end of the week when the ${item.button} button is clicked`, () => {
```
Comment (Thebarda) [inline centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx:385]:
```suggestion

    it(`checks that the last day must match with the end of the week when the ${item.button} button is clicked`, () => {
```

File: centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx [hunk 18]
```diff
@@ -0,0 +1,471 @@
...
+              formatString: 'dd'
+            })
+          );
+          daysArray.forEach((day) => cy.contains(day.toUpperCase()));
+          const currentMonth = Object.keys(element)[0];
+          getPreviousNextMonth({ currentMonth, labelButton: item.button });
+        });
+      });
+    });
+    it(`displays the calendar for the timezone ${item.timezone} when the ${item.button} button is clicked`, () => {
```
Comment (Thebarda) [inline centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx:456]:
```suggestion

    it(`displays the calendar for the timezone ${item.timezone} when the ${item.button} button is clicked`, () => {
```

File: centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx [hunk 19]
```diff
@@ -0,0 +1,471 @@
...
+              />
+            </LocalizationProvider>
+          )
+        });
+      });
+    });
+
+    it('checks input calendar value contains correct date', () => {
+      const { result } = renderHook(() => useLocaleDateTimeFormat());
+      act(() => {
```
Comment (Thebarda) [inline centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx:315]: Is it mandatory to wrap it inside an act?
act is meant to be used when you update a state or something

File: centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx [hunk 20]
```diff
@@ -0,0 +1,463 @@
...
+    January: {
+      date: '2023-01-01T12:59:41.041Z',
+      firstDay: { day: days[0], indexDayInRowWeek: 0, value: 1 },
+      lastDay: { day: days[2], indexDayInRowWeek: 2, value: 31 },
+      numberWeeks: 5
+    }
+  }
+];
+
+const dataTest = [
```
Comment (bdauria) [inline centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx:235]:
```suggestion
const testData = [
```

File: centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx [hunk 21]
```diff
@@ -0,0 +1,463 @@
...
+enum ButtonCalendar {
+  NMONTH = 'Next month',
+  PMONTH = 'Previous month'
+}
+interface GetPreviousNextMonth {
+  currentMonth: string;
+  labelButton: string;
+}
+
+const getPreviousNextMonth = ({
```
Comment (bdauria) [inline centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx:259]: I would prefer to have to functions: clickOnPreviousMonth and clickOnNextMonth

File: centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx [hunk 22]
```diff
@@ -0,0 +1,463 @@
...
+};
+
+const changeDate = (): void => undefined;
+
+const setStart = undefined;
+const checkIfDuplicateExists = (arr: Array<unknown>): boolean => {
+  return new Set(arr).size !== arr.length;
+};
+
+dataTest.forEach((item) =>
```
Comment (bdauria) [inline centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx:282]:
```suggestion
testData.forEach((item) =>
```

File: centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx [hunk 23]
```diff
@@ -0,0 +1,463 @@
...
+
+      const dateInput = format({
+        date: dayjs(item.initialDate).tz(item.timezone),
+        formatString: 'L hh:mm A'
+      });
+
+      cy.get('input').should('have.value', dateInput);
+    });
+
+    it(`checks number of days in current month when the ${item.button} button is clicked`, () => {
```
Comment (bdauria) [inline centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx:321]:
```suggestion
    it(`displays the correct number of days for the current month when the ${item.button} button is clicked`, () => {
```

File: centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx [hunk 24]
```diff
@@ -0,0 +1,463 @@
...
+        cy.get('[role="rowgroup"]').children().as('listWeeks');
+        cy.get('@listWeeks').children('button').as('days');
+        cy.get('@days').should('have.length', lastDay.value);
+
+        const currentMonth = Object.keys(element)[0];
+        getPreviousNextMonth({ currentMonth, labelButton: item.button });
+      });
+    });
+
+    it(`does not duplicate days in each month of the year, when the ${item.button} button is clicked`, () => {
```
Comment (bdauria) [inline centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx:335]:
```suggestion
    it(`does not duplicate days in any of the month in the year when the ${item.button} button is clicked`, () => {
```

File: centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx [hunk 25]
```diff
@@ -0,0 +1,463 @@
...
+          .eq(firstDay.indexDayInRowWeek)
+          .as('firstDayInWeek');
+        cy.get('@firstDayInWeek').contains(firstDay.value);
+
+        const currentMonth = Object.keys(element)[0];
+        getPreviousNextMonth({ currentMonth, labelButton: item.button });
+      });
+    });
+
+    it(`checks that the last day must match with the end of the week when the ${item.button} button is clicked`, () => {
```
Comment (bdauria) [inline centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx:380]:
```suggestion
    it(`displays the last day as the end of the week when the ${item.button} button is clicked`, () => {
```

File: centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx [hunk 26]
```diff
@@ -0,0 +1,463 @@
...
+        cy.get('@listDaysInLastWeek')
+          .eq(lastDay.indexDayInRowWeek)
+          .as('lastDayInWeek');
+        cy.get('@lastDayInWeek').contains(lastDay.value);
+        const currentMonth = Object.keys(element)[0];
+        getPreviousNextMonth({ currentMonth, labelButton: item.button });
+      });
+    });
+
+    it(`display the month and the year in the calendar's header ,when the ${item.button} button is clicked`, () => {
```
Comment (bdauria) [inline centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx:406]:
```suggestion
    it(`displays the month and the year in the calendar's header when the ${item.button} button is clicked`, () => {
```

File: centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx [hunk 27]
```diff
@@ -0,0 +1,463 @@
...
+          date: dayjs(date).tz(item.timezone),
+          formatString: 'MMMM YYYY'
+        });
+        cy.contains(monthAndYear);
+        const currentMonth = Object.keys(element)[0];
+        getPreviousNextMonth({ currentMonth, labelButton: item.button });
+      });
+    });
+
+    it(`displays the correct name of days on calendar's header, when the ${item.button} button is clicked`, () => {
```
Comment (bdauria) [inline centreon/www/front_src/src/Resources/Graph/Performance/TimePeriods/index.cypress.spec.tsx:424]:
```suggestion
    it(`displays the correct day name on calendar's header when the ${item.button} button is clicked`, () => {
```

**Comment by bdauria** [comment]
@Noha-ElAbrouki I don't think the cypress spec belongs to the PerformanceGraph directory

**Review by bdauria — APPROVED** [review]

**Review by Thebarda — APPROVED** [review]
